### ISSUE TITLE
Site duplication still an issue

### ISSUE THREAD

**Issue by jlbruce** [USGS-WiM/SiGLDMS#154]
Bad news, site duplication is still really messed up. The problem fields seem to be Resource Component (RC), Media (M), Sampling Frequency (SF), and Parameters. I tried duplicating a few sites and got all kinds of mistakes, both display and content, and got really confused. So I started a new site and here's what happened...

**FIRST SITE**
- Initial site creation: kept the info I entered. After site was initially saved, RC, M, and SF content didn't display in list. Parameters did.
- Site duplication:
  _Actual data:_ Dropped all three RC values and substituted three of its own. Dropped one M value, kept 2, added 1. Dropped 3 S values, added 1. Kept all parameter values, but added 4 extras.
  _Display of duplicated site:_ RC, M, and SF content didn't display in list. Parameters did.
- Note: some of the values it inserted might have come from previously duplicated sites (from my initial, undocumented duplication trials)

**SECOND SITE**
- Initial site creation: kept the info I entered. All fields displayed correctly
- Site duplication:
  _Actual data:_ Kept both my initial RC values, added the same three values it added to the first duplicated site (not the values I had initially entered for the first site). Kept both my initial M values, added the same three values it added to the first duplicated site. Kept my initial SF values, added the same value it added to the first duplicated site. Kept my original parameters plus all the parameters from the first duplicated site (my original plus the extras).
  _Display of duplicated site:_ RC, M, and SF content didn't display in list. Parameters did.

Another note: after clicking onto the Publications tab and then clicking back to the Sites tab (where it reloads the site list), the empty RC, M, and F fields were now populated. However, Site 1 Duplicate displayed some content multiple times ("Triennial, Triennial") - not sure how THAT happens since you can't pick the term more than once - in the RC, F, (but not M) and Parameters (only Biological, not the others). Site 2 Duplicate did not display any multiples.

So, clearly there's some display issues (although this erroneous behavior is erratic) and also some carrying over of values from previously duplicated sites. For some duplications, it seems to replace values with the carry over, in other cases, it just keeps adding them.

What a mess. Good luck!

**Comment by troddick** [comment]
Thought I had this fixed. Turns out if you just cancel out of the modal that comes up once you duplicate a site (rather than hitting save, which is only useful if you actually make a change within the modal), everything shows up ok. 
I then turned off that step, so the duplication happens, and the list updates just fine. I can leave it this way, and if the user wants to edit the newly duplicated site, they click edit button. 
But then, I clicked on it and hit save without making any changes, and the media, resources and frequences were gone. So, it's a save issue. If no changes are happening to those things (M, R, F), then they aren't getting passed back to the list view from the edit view (even though they are still attached). I have to fix this irregardless if I remove the modal from popping up with the newly duplicated site or not. But I think it looks better without the modal popping up, but I leave that open to you. 
I'll look into the repeating "Triennial, Triennial" issue as well.

**Comment by jlbruce** [comment]
I'm fine with not having a modal show up when you duplicate a site, although in case the user wants to edit it immediately, it would be nice for the duplicate site to either: 1) the duplicate site row to be highlighted in some way (until another action is performed) and/or 2) have the duplicate site row show up directly below the original, regardless of the sort order (until another sort order is initiated).

One question, though... your discussion above didn't mention the assignment of unselected data (likely from previous duplications, which is happening at the actual site data level, not just the site list display). Will the fixes you talk about above also solve this issue?

**Comment by troddick** [comment]
I'm not getting these same errors. Did you say you did the hard refresh (totally clearing your cache)? I've noticed Chrome really likes hanging on to the javascript files. Even when I hard refresh (Ctrl+F5) and even close and reopen my browser. I usually end up going back and forth between Chrome and FF for debugging to deal with this issue. 

I'm going to update the dev server's code right now with the updates. I think I fixed the edit site and duplicate site issues. Maybe give that a test (after you've deleted all your cookies and cache) and let me know if there is still weirdness going on. If so, we can do a hangout tomorrow to take a look together.

**Comment by jlbruce** [comment]
I did not try a hard refresh or cache clear **during** my site duplication testing (but I have been doing a lot of them lately for another project, so I know everything SiGL was fresh **before** I started testing).

I'll do a hard clearing again tomorrow and retest the duplication issues and let you know. Thanks!

**Comment by jlbruce** [comment]
Ok, tested again. I cleared my cache before doing anything.

I added a new site C, and duplicated it, and all was successful.

I added a second new site D, and duplicated it, and all displayed successfully in the list. But when I went it to edit it (check the content), it had picked up the R, M, F and Params from C and added them to the list.

I went back in to check the content of site C's duplicate after that, and the original content was still fine (no duplicates). Only the second duplicate had been affected.

**Comment by jlbruce** [comment]
UPDATE: I realized that I hadn't deleted my cookies (only cache and download/browsing history), so I re-cleared and tried again. It worked exactly like the process above (only the second duplication was affected), but I tested a couple of things...

Before going into editing mode for the second site duplicate (site F), I clicked into Pubs and then came back. The site list for the site F duplicate now had the extra terms, even though I hadn't gone into the edit mode, so it must have added them during the site creation, not attached them because of the edit mode.

Also, I went back and made a second duplicate of my first site (site E), and it did not add the extra variables (either in the list view or when going into edit). For some reason, it didn't carry them BACK to the first site from the second site. ???

**Comment by troddick** [comment]
OK, I was able to duplicate your issue with creating Site C, duplicating, then creating site D, duplicating and seeing site C's stuff on site D when trying to edit. 
I'm sorry this is taking a few iterations to fix. The logic on this section is a little detailed. Hopefully I've fixed it this time. After fixing it, I did go to Pubs and then back to sites, and all still looked good. I did not see the extra terms, but may have not done it the exact same way you did it, or I did fix this with the other fix.
Please try again (hard clear your cache first) and let me know. Keeping my fingers crossed. FYI: I'm out until Thursday.
Thanks

**Comment by jlbruce** [comment]
Good news! I tested the duplication process and all seems to be going well. I didn't see the DMS adding or messing with the data in any way. And the display issues also seem almost entirely resolved, except for one thing...

I did my standard "create site 1, duplicate 1, create site 2, duplicate 2" process and then checked all the sites. When just creating and editing sites, all was fine. When I clicked into the Pubs tab and came back to Sites, the values of the RC, M, and F fields were reordered - only for Duplicate 2. So while Original 2 would display "Harbors, Embayment", Duplicate 2 would display "Embayment, Harbors". This isn't a terribly big deal, except that it makes site comparison harder. But at least the data is preserved.

Regardless, excellent work!!!
